Rubberduck.Parsing/VBA/VBACodeStringParser.cs

```diff
@@ -0,0 +1,112 @@
+        /// Parse a given string representing a module code. The code passed in must be
+        /// a valid module body, containing valid declarations or complete procedures. 
+        /// Code snippets are not valid. 
+        /// </summary>
+        /// <param name="moduleName">For logging purpose, provide descritpive name for where the virtual module is used</param>
+        /// <param name="moduleCodeString">A string containing valid module body.</param>
+        /// <param name="mode">Indicates what parser mode to use. By default we use SLL fallbacking to LL. When mode is explicitly specified, only that mode will be used with no fallback.</param>
+        public VBACodeStringParser(string moduleName, string moduleCodeString, ParserMode mode = ParserMode.Default)
+        {
+            var tokenStreamProvider = new SimpleVBAModuleTokenStreamProvider();
+            _tokenStream = tokenStreamProvider.Tokens(moduleCodeString);
+            _parser = new VBAParser(_tokenStream);
+            _parser.AddErrorListener(new MainParseExceptionErrorListener(moduleName, ParsePass.CodePanePass));
+            _moduleName = moduleName;
+            _mode = mode;
+        }
+
+        public (IParseTree parseTree, TokenStreamRewriter rewriter) Parse()
+        {
+            var parseTree = _mode == ParserMode.Default ? ParseInternal(_moduleName) : ParseInternal(_mode);
+            var rewriter = new TokenStreamRewriter(_tokenStream);
+
+            return (parseTree, rewriter);
+        }
+
+        private IParseTree ParseInternal(string moduleName)
+        {
+            try
+            {
+                return ParseInternal(ParserMode.Sll);
+            }
+            catch (ParsePassSyntaxErrorException syntaxErrorException)
+            {
+                var parsePassText = syntaxErrorException.ParsePass == ParsePass.CodePanePass
+                    ? "code pane"
+                    : "exported";
+                var message = $"SLL mode failed while parsing the {parsePassText} version of module {moduleName} at symbol {syntaxErrorException.OffendingSymbol.Text} at L{syntaxErrorException.LineNumber}C{syntaxErrorException.Position}. Retrying using LL.";
+                LogAndReset(message, syntaxErrorException);
+                return ParseInternal(ParserMode.Ll);
+            }
+            catch (Exception exception)
+            {
+                var message = $"SLL mode failed while parsing module {moduleName}. Retrying using LL.";
+                LogAndReset(message, exception);
+                return ParseInternal(ParserMode.Ll);
+            }
+        }
+
+        private IParseTree ParseInternal(ParserMode mode)
+        {
+            if (mode == ParserMode.Ll)
+            {
+                _parser.Interpreter.PredictionMode = PredictionMode.Ll;
+            }
+            else
+            {
+                _parser.Interpreter.PredictionMode = PredictionMode.Sll;
+            }
+            return _parser.startRule();
```

Comment:
Not sure how we can do this, but being able to just tokenize a string by specifying the entry rule would be awesome.

Perhaps with a *member selector* `Expression` against `VBAParser`? e.g.:

    var result = p.Parse("ForLoopBlock", code, ParserMode.Sll, p => p.forNextStmt);

That way the calling code could easily tokenize a specific block of code using the appropriate entry point, instead of making the parser figure out what rule matches the input.

Reply:
...not sure reflecting on `VBAParser` members is less expensive than letting the parser figure out what rule matches the input though.

Reply:
That's good idea. I would prefer an explicit context from the calling code than letting parser guess, that way the responsibility is on the calling code and hopefully makes for readable code where we really want to deal with fragments. I'll see if this is possible to get it to parse from a particular context rather than at the top level.

Reply:
@bclothier it's definitely possible, but then you get to deal with `Expression` trees meta-code and reflection to invoke the specified rule. It's absolutely fun, but can be mind-wrecking :smile: